Add initial module version

irm-codebase opened the pull request:
Includes:
- Country shapes from GADM and Overture Maps
- EEZ from Marine regions

Data is harmonized into a single schema, and then combined.
Contested maritime claims are removed by default to avoid thorny research.

Marine regions is the best source for country borders, so it is given priority during the merge by 'clipping' the other shapes to it.

irm-codebase commented:
Here is an initial example showing Japan, South Korea and Taiwan from different sources.
It's the integration test, for now. I might choose something different if it is too heavy for github's CI.

![image](https://github.com/user-attachments/assets/8e9b8c53-9e96-4359-9fa0-88fa682d154b)

brynpickering commented on workflow/scripts/standardise_country_gadm.py:
```diff
@@ -0,0 +1,51 @@
+        subtype (str): regional definition in the parent dataset (0, 1, 2).
+        output_path (str): path to output file.
+    """
+    gdf = gpd.read_parquet(input_path)
+    gdf["country_id"] = gdf["GID_0"]
+    gdf["class"] = "land"
+    gdf["parent"] = "gadm"
+    gdf["parent_subtype"] = subtype
+    gdf["parent_id"] = gdf[f"GID_{subtype}"]
+    gdf["parent_name"] = gdf[f"NAME_{subtype}"]
+    gdf["shape_id"] = gdf["parent_id"].apply(lambda x: f"{country_id}_gadm_{x}")
+    gdf = gdf[
+        [
+            "shape_id",
+            "country_id",
+            "class",
+            "geometry",
+            "parent",
+            "parent_subtype",
+            "parent_id",
+            "parent_name",
+        ]
+    ]
+    gdf.to_parquet(output_path)
```
I would have preference for creating a new geodataframe rather than appending columns to the current one and then selecting those new columns to strip out the unwanted ones. That is, something like:

```py
new_gdf = gpd.GeoDataFrame(
    {
        "country_id": gdf["GID_0"],
        "class": "land",
        ....
    }
)
```

irm-codebase replied:
Good suggestion!
yeah, that's a better practice.

brynpickering commented on workflow/internal/config.schema.yaml:
I assume we're not going to move towards using pydantic for the config schema...? I realise this is a data module template question, but one I'm raising here now that I see a JSON schema being defined in the wild...

irm-codebase replied:
We cannot, unfortunately, because this is core `snakemake` functionality.
Unless `snakemake` moves to pydantic, we are stuck with these for validating module configuration

brynpickering replied:
Does snakemake validate automatically if there is a schema? If we are already validating manually, we could just give it a json schema dumped from a pydantic model...?

irm-codebase replied:
it validates via `validadate(...)` in the `Snakefile`.

```python
# Validate the configuration using the schema file.
validate(config, workflow.source_path("internal/config.schema.yaml"))
```

Regarding using `pydantic`: unfortunately no. 

Snakemake code outside of rules runs on the 'main' environment used by a user to run their workflow and our modules. `pydantic` is not a core dependency in python or in snakemake, meaning that we would have to 'inject' it. I already tried doing this, and it is very messy.

https://snakemake.readthedocs.io/en/stable/snakefiles/deployment.html#global-workflow-dependencies

It is way safer to just rely on `snakemake`'s validation functions here.

brynpickering reviewed (commented):
Some initial thoughts!

> Marine regions is the best source for country borders, so it is given priority during the merge by 'clipping' the other shapes to it.

By this you mean that the coastal borders are clipped to the marine regions, right? Might be worth clarifying this order of priority in the README. 

Also, what happens when the land border doesn't reach the marine region? The current clipping will just ensure there isn't overlap, but won't gap fill?

irm-codebase commented on pixi.toml:
```diff
@@ -20,6 +20,13 @@ conda = ">=25.0.0"
 snakemake-minimal = ">=8.29.0"
 cffconvert = ">=2.0.0"
 clio-tools = ">=2025.03.03"
+ipykernel = ">=6.29.5"
```
Sounds reasonable. Could you open an issue in the template?

irm-codebase commented on workflow/scripts/download_country_overture.py:
```diff
@@ -0,0 +1,69 @@
+    for extension in ["spatial", "httpfs"]:
+        connection.install_extension(extension)
+        connection.load_extension(extension)
+    connection.sql("SET s3_region='us-west-2'")
```
Nope. The download is still fairly fast though.

irm-codebase commented on workflow/scripts/download_marine_eez_area.py:
```diff
@@ -0,0 +1,94 @@
+    snakemake: Any
+
+MARINE_URL = "https://www.marineregions.org/download_file.php"
+FILE = "World_EEZ_v12_20231025_gpkg"
```
Making this configurable could lead to a lot of user confusion, and not really protect us against anything.
If a new version becomes available, there are no guarantees that it will be backwards compatible.

I also decided to place this here instead of in `internal/` because it is not really shared between rules, so the added overhead did not seem worth it.

irm-codebase commented on workflow/scripts/download_marine_eez_area.py:
```diff
@@ -0,0 +1,94 @@
+            "organisation": org,
+            "email": f"{name}@{org}.org",
+            "country": "Netherlands",
+            "user_category": "academia",
```
I guess... but this does not strike me as an important thing to make configurable.
It's just a download request.

brynpickering replied:
So long as they don't impose a different license depending on use...

irm-codebase replied:
They use CC-BY: https://www.marineregions.org/disclaimer.php

irm-codebase commented on INTERFACE.yaml:
There is no easy way of doing this check, so no.
It's one of the things that has to be kept up to date by devs, but it's fairly minimal.

irm-codebase commented on README.md:
```diff
@@ -4,14 +4,19 @@ A data module to create arbitrary regional boundaries for energy systems modelli
 
 A modular `snakemake` workflow built for [`clio`](https://clio.readthedocs.io/) data modules.
 
+## Using this module
```
Agreed. This should be in the template.

irm-codebase commented on workflow/envs/shape.yaml:
```diff
@@ -0,0 +1,11 @@
+  - conda-forge
+  - nodefaults
+dependencies:
+  - duckdb = 1.2.0
+  - geopandas = 1.0.1
+  - ipdb = 0.13.13
+  - pandera-geopandas = 0.23.1
+  - pyarrow = 19.0.1
+  - pycountry = 24.6.1
```
pixi deps are never seen by `snakemake` rules because they generate independent `conda` environments. They are exclusively for testing stuff outside of snakemake. They do not matter as far as the module is concerned.

I could even remove them from the `pixi.toml` and the module would still run.

brynpickering replied:
Then why have them as pixi deps at all? If they are meant to only ever be used in the isolated environment then they shouldn't crop up in the pixi deps.

irm-codebase replied:
Agreed!
I thought they might be useful for debugging in notebooks, but it does not really make sense.
Removed!

irm-codebase commented on workflow/scripts/_schema.py:
```diff
@@ -0,0 +1,18 @@
+
+import pandera as pa
+
+schema = pa.DataFrameSchema(
+    {
+        "shape_id": pa.Column(str),
+        "country_id": pa.Column(str),
+        "class": pa.Column(str, pa.Check.isin(["land", "maritime"])),
+        "geometry": pa.Column("geometry"),
+        "parent": pa.Column(str, pa.Check.isin(["gadm", "overture", "marineregions"])),
+        "parent_subtype": pa.Column(str),
+        "parent_id": pa.Column(str),
+        "parent_name": pa.Column(str),
+    },
+    coerce=True,
+    strict=True,
+)
```
Not sure if that would allow column specific checks...
Can you point to an example?

brynpickering replied:
Ignore me, even using `columns={..}` doesn't let you avoid having to wrap types in `pa.Column`

irm-codebase commented:
> Neither do I... But perhaps just being clear about it in the description is sufficient (since nobody else is doing it anyway, so it doesn't make this module any "worse" than other options!).

Fair!
I've updated the README. I also modified the scripts a bit to ensure that landmasses are still 'clipped' for all territories.
Previously, contested EEZs would not be clipped, which gives amorphous 'blobs' in some cases (see https://github.com/OvertureMaps/data/issues/346)

Now, the landmass will always be correct, but contested EEZs will not be attributed to either party.

brynpickering commented on .copier-answers.yaml:
```diff
@@ -1,13 +1,14 @@
 # DO NOT MANUALLY MODIFY THIS FILE
-_commit: 977c0ba
-_src_path: ./
+_commit: v0.1.0
```
not going for calver? ;)

irm-codebase replied:
That's a data module template thing ;^)
I went for semver for that one to avoid confusion between clio versions and template versions. Perhaps they should match...

This field should be auto-updated when updating from the template regardless.

brynpickering commented on workflow/scripts/build_combined_area.py:
```diff
@@ -5,19 +5,74 @@
+def _combine_shapes(
+    country_files: list[str], marine_file: str, geographic_crs: str
+) -> gpd.GeoDataFrame:
+    """Merge each all countries and maritime boundaries into one file.
```
each or all?

brynpickering commented on workflow/scripts/build_combined_area.py:
```diff
@@ -5,19 +5,74 @@
+        # Ensure the resulting shape has no bowties
+        if not new_geometry.is_valid:
+            new_geometry = new_geometry.buffer(0)
+            assert new_geometry.is_valid
```
add a message if assertion fails?

brynpickering commented on workflow/internal/config.schema.yaml:
```diff
@@ -2,6 +2,26 @@ $schema: "https://json-schema.org/draft/2020-12/schema"
+  buffer:
+    description: >
+      Buffer boundary between shapes. Unit will be the one used by the projected CRS.
+      If zero, no buffer will be applied, which may lead to overlaps between shapes.
```
may lead to "overlaps between shapes" or "gaps between shapes"? I assume the buffer is used to fill possible gaps, which are then clipped against the actual boundaries?

irm-codebase replied:
The algorithm in its current form cannot correct gaps between shapes, only overlaps, so I only mention that here.

Saying otherwise might be misleading, I think.

brynpickering commented on workflow/scripts/build_combined_area.py:
```diff
@@ -28,21 +83,40 @@ def build_combined_area(
+            )
             combined = pd.concat(
-                [combined, country_land, country_marine], ignore_index=True
+                [combined, country_land, uncontested], ignore_index=True
```
this should pergaps be optional? I.e. users can get contested EEZs if they need them for their project (with a "contested" flag)?

irm-codebase replied:
This omission is half 'ethics' and half avoiding complex implementation.

The problem of turning this off is that we have to choose which party owns a border, which implies choosing a country_id etc. Adding an extra column for "contested" is also an issue for landmasses, as those would be left without information.

I decided to go for this instead of providing that information half-way.

irm-codebase replied:
In the future it'd be nice to give this info, but for now this seemed more pragmatic.
